pkg/ottl/parser_test.go

```diff
@@ -1462,7 +1462,7 @@ func Test_parseStatement(t *testing.T) {
 		{`set() where true and 1 == int() `, true},
 		{`set() where false or 1 == int() `, true},
 		{`set(foo.attributes["bar"].cat, "dog")`, false},
-		{`set(foo.attributes["animal"], "dog") where animal == "cat"`, false},
+		{`set(set = foo.attributes["animal"], val = "dog") where animal == "cat"`, false},
```

Comment:
This test just checks that the parsing library can parse the grammar, but I've made an edit to `replace_pattern` to use it as the example function for this testing. I've tested the following statements:
- `replace_pattern(attributes["device"], ".*", "$$1")`
- `replace_pattern(attributes["device"], ".*", "$$1", function = "SHA256")`
- `replace_pattern(attributes["device"], ".*", "$$1", "SHA256")`

Reply:
Could we also add a test that demonstrates passing in an optional param and otherwise? I think that test would be in `function_test.go`

Reply:
I'm trying to keep tests light since this is PR is still undergoing heavy revisions, but I'll make sure to add that before I mark the PR as ready for a proper review.